yellowbrick/cluster/silhouette.py

```diff
@@ -135,18 +135,30 @@ def fit(self, X, y=None, **kwargs):
         # NOTE: Probably this would be better in score, but the standard score
         # is a little different and I'm not sure how it's used.
 
-        if not check_fitted(self.estimator, is_fitted_by=self.is_fitted):
-            # Fit the wrapped estimator
-            self.estimator.fit(X, y, **kwargs)
+        # if estimator is fitted AND has attribute
+        if check_fitted(self.estimator, is_fitted_by=self.is_fitted) and hasattr(self.estimator, "predict"):
+            labels = self.estimator.predict(X)
+        else:  # if estimator is NOT fitted, OR estimator does NOT implement predict()
+            labels = self.estimator.fit_predict(X, y, **kwargs)
 
         # Get the properties of the dataset
         self.n_samples_ = X.shape[0]
-        self.n_clusters_ = self.estimator.n_clusters
+
+        if hasattr(self.estimator, "n_clusters"):
+            self.n_clusters_ = self.estimator.n_clusters
+        else:
+            self.n_clusters_ = len(set(labels)) - (1 if -1 in labels else 0)
+
+        if hasattr(self.estimator, "metric"):
+            metric = self.estimator.metric
+        elif hasattr(self.estimator, "affinity"):
+            metric = self.estimator.affinity
```

Comment:
@lwgray this is where the error is occurring for SpectralClustering - `SpectralClustering` does have an attribute `affinity` which is used to compute the adjacency matrix between instances. For spectral clustering the attribute `affinity` is not a distance metric and defaults to "rbf" -- which is why that test is failing. 

@stergion what model prompted you to add this metric selector?

Reply:
it was for the AffinityPropagation, AgglomerativeClustering, FeatureAgglomeration.

Since sklearn version 1.2, `affinity` is deprecated in AgglomerativeClustering and FeatureAgglomeration 
and `metric` is used instead, like the other clustering algorithms.

AffinityPropagation was not updated, it still uses `affinity`. Although, AffinityPropagation uses the negative 
squared euclidean distance between points, when `affinity='euclidean'`